apps/miiverse-api/src/services/internal/middleware/check-user-account.ts

```diff
@@ -0,0 +1,23 @@
+import { errors } from '@/services/internal/errors';
+import type express from 'express';
+
+/**
+ * Checks the user account is valid for this request (not banned, setup complete, etc.)
+ */
+async function checkUserAccount(request: express.Request, response: express.Response, next: express.NextFunction): Promise<void> {
+	const account = response.locals.account;
+	if (account === null) {
+		// Guest access
+		return next();
+	}
+
+	if (account.pnid.accessLevel < 0) {
+		throw new errors.unauthorized('Account has been banned');
+	}
+
+	// TODO account settings, temp bans, etc.
+
+	return next();
+}
```

Comment:
Maybe we should more action driven middlewares like `canUserWrite`.
Cuz banned users can still read posts, right?

Reply:
I'm pretty sure banned users can't even log in and have to use the anonymous access. I do still have a bit of action-driven stuff on the routes itself:
```js
postsRouter.get('/:post_id', authGuest, //...
```
This would allow guest (and I guess I could special-case banned too) reading of posts; but we'd have something else like:
```js
postsRouter.post('/:post_id/empathy', authUser, //...
```
...so only logged-in users can Yeah.

Reply:
How about these middleware?
```ts
router.use(populateAuth); // token handling
router.use(authAccess); // prevent unauthorized users from accessing

router.get('/:id', guards.guest, ...);
router.get('/:id', guards.user, ...);
router.get('/:id', guards.mod, ...);
```